Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java

```diff
@@ -0,0 +1,120 @@
+        destinationPid = pidFactory.createPIDManager("DriveToDestination", 0.1, 0, 0, 0, 1, -1, 3, 1, 0.5);
+        
+        finalHeadingEffectDistance = propMan.createPersistentProperty("DriveHeuristics heading effect distance", 12.0);
+        driveToLineEffectDistance = propMan.createPersistentProperty("DriveHeuristics line effect distance", 12.0);
+    }
+
+    @Override
+    public void initialize() {
+        FieldPose currentPoint = poseSubsystem.getCurrentFieldPose();
+        
+        // for testing, let's just go ahead a few feet while turning left 90 degrees
+        double heading = currentPoint.getHeading().getValue();
+        
+        // x and y to add
+        double yDelta = (Math.sin(Math.toRadians(heading)) * 60) + (Math.sin(Math.toRadians(heading+90)) * 60);
+        double xDelta = (Math.cos(Math.toRadians(heading)) * 60) + (Math.cos(Math.toRadians(heading+90)) * 60);
+        
+        goalPoint = new FieldPose(
+                new XYPair(
+                        currentPoint.getPoint().x + xDelta,
+                        currentPoint.getPoint().y + yDelta),
+                new ContiguousHeading(heading + 90));
+    }
+
+    /**
+     * There are three "forces" at work for this approach.
+     * 1) Align to the goal heading. This force grows the closer we are to the line of travel
+     * 2) Align towards the line of travel. This force grows the further we are from the line of travel
+     * 3) Drive towards the final point. This is a constant force.
+     * 
+     * Each of these is basically a PID (optionally combined with a scaling factor)
+     * 
+     * There are some small quirks as well:
+     *  If the final point is "behind" the robot, the robot should drive backwards, not forwards. This 
+     *  means that force (2) can sometimes use a heading towards the line + 180 degrees.
+     */
+    @Override
+    public void execute() {       
+        FieldPose currentPose = poseSubsystem.getCurrentFieldPose();
+        
+        double distanceToLine = goalPoint.getDistanceToLineFromPoint(currentPose.getPoint());
+        double yDisplacement = goalPoint.getPointRelativeYDisplacementFromLine(currentPose);
+        
+        // once we are within 6 inches, this begins to get important
+        double finalHeadingScalingFactor = finalHeadingEffectDistance.get() / (12 + finalHeadingEffectDistance.get());
+        finalHeadingScalingFactor = finalHeadingScalingFactor > 1 ? 1: finalHeadingScalingFactor;
+        
+        double alignTowardsLineScalingFactor = distanceToLine / driveToLineEffectDistance.get();
+        alignTowardsLineScalingFactor = alignTowardsLineScalingFactor > 1 ? 1 : alignTowardsLineScalingFactor;
+        
+        ContiguousHeading alignTowardsLineHeading = 
+                goalPoint.getLinePerpendicularToPoint(currentPose.getPoint()).getHeading();
+        alignTowardsLineHeading = yDisplacement > 0 ? alignTowardsLineHeading : alignTowardsLineHeading.shiftValue(180);
```

Comment:
I think it's mostly a moot point - if `yDisplacement` is at zero, it means (at worst) that you're oscillating between two opposite points on a circle, which are essentially equivalent in terms of how they will make the robot turn to align towards the line.